topics/error-handling.md

```diff
@@ -0,0 +1,168 @@
+# Error handling guidelines
+
+Almost all errors in Cartridge follow `return nil, err` style, where
+`err` is an error object produced by
+[`errors`](https://github.com/tarantool/errors) module. Cartridge
+doesn't raise errors except for bugs and functions contracts mismatch.
+Developing new roles should follow these guidelines as well.
+
+## Error objects in Lua
+
+Error classes help to locate the problem's source. For this purpose, an
+error object contains its class, stack traceback and a message.
+
+```lua
+local errors = require('errors')
+local DangerousError = errors.new_class("DangerousError")
+
+local function some_fancy_function()
+
+    local something_bad_happens = true
+
+    if something_bad_happens then
+        return nil, DangerousError:new("Oh boy")
+    end
+
+    return "success" -- not reachable due to the error
+end
+
+print(some_fancy_function())
+```
+
+```
+nil	DangerousError: Oh boy
+stack traceback:
+	test.lua:9: in function 'some_fancy_function'
+	test.lua:15: in main chunk
+```
+
+For uniform error handling `errors` provides `:pcall` API:
+
+```lua
+local ret, err = DangerousError:pcall(some_fancy_function)
+print(ret, err)
+```
+
+```text
+nil	DangerousError: Oh boy
+stack traceback:
+	test.lua:9: in function <test.lua:4>
+	[C]: in function 'xpcall'
+	.rocks/share/tarantool/errors.lua:139: in function 'pcall'
+	test.lua:15: in main chunk
+```
+
+```lua
+print(DangerousError:pcall(error, 'what could possibly go wrong?'))
+```
+
+```text
+nil	DangerousError: what could possibly go wrong?
+stack traceback:
+	[C]: in function 'xpcall'
+	.rocks/share/tarantool/errors.lua:139: in function 'pcall'
+	test.lua:15: in main chunk
+```
+
+For `errors.pcall` there is no difference between `return nil, err` and
+`error()` approaches.
+
+Note that `errors.pcall` API differs from vanilla lua
+[`pcall`](https://www.lua.org/pil/8.4.html). Instead of `true` former
+returns values returned from the call. If there is error, it returns
+`nil` instead of `false`, plus error message.
+
+Remote `net.box` calls don't keep stack trace from the remote. In that
+case `errors.netbox_eval` comes to the rescue. It will find stack trace
+from local and remote hosts and restore metatables.
+
+```
```

Comment:
I think we can add "lua" here.

Try to look at result page - one pieces of code/examples looks quite poor.